frontend/app/controllers/spree/checkout_controller.rb

```diff
@@ -25,33 +25,62 @@ class CheckoutController < Spree::StoreController
 
     # Updates the order and advances to the next state (when possible.)
     def update
-      if OrderUpdateAttributes.new(@order, update_params, request_env: request.headers.env).apply
-        @order.temporary_address = !params[:save_user_address]
-        success = if @order.state == 'confirm'
-          @order.complete
-        else
-          @order.next
-        end
-        if !success
-          flash[:error] = @order.errors.full_messages.join("\n")
-          redirect_to(checkout_state_path(@order.state)) && return
-        end
+      if update_order
+
+        assign_temp_address
+
+        redirect_on_failure and return unless transition_forward!
```

Comment:
It is definitely more readable and more accessible even for people coming from other languages.